src/redisearch_rs/inverted_index/src/lib.rs

```diff
@@ -497,11 +543,147 @@ impl<E: Encoder> InvertedIndex<E> {
     }
 }
 
+/// Result of scanning the index for garbage collection
+#[derive(Debug, Eq, PartialEq)]
+pub struct GcScanDelta {
+    /// The index of the last block in the index at the time of the scan. This is used to ensure
+    /// that the index has not changed since the scan was performed.
+    last_block_idx: usize,
+
+    /// The number of entries in the last block at the time of the scan. This is used to ensure
+    /// that the index has not changed since the scan was performed.
+    last_block_num_entries: usize,
+
+    /// The results of the scan for each block that needs to be repaired or deleted.
+    deltas: Vec<BlockGcScanResult>,
+}
+
+/// Result of scanning a block for garbage collection
+#[derive(Debug, Eq, PartialEq)]
+pub struct BlockGcScanResult {
+    /// The index of the block in the inverted index
+    index: usize,
+
+    /// The type of repair needed for this block
+    repair: RepairType,
+}
+
+/// Information about the result of applying a garbage collection scan to the index
+#[derive(Debug, Eq, PartialEq)]
+pub struct GcApplyInfo {
+    /// The number of bytes that were freed
+    bytes_freed: usize,
+
+    /// The number of bytes that were allocated
+    bytes_allocated: usize,
+
+    /// The number of entries that were removed from the index
+    entries_removed: usize,
+
+    /// The number of blocks that were ignored because the index changed since the scan was performed
+    blocks_ignored: usize,
+}
+
 impl<E: Encoder + DecodedBy> InvertedIndex<E> {
     /// Create a new [`IndexReader`] for this inverted index.
     pub fn reader(&self) -> IndexReader<'_, E, E::Decoder> {
         IndexReader::new(self)
     }
+
+    /// Scan the index for blocks that can be garbage collected. A block can be garbage collected
+    /// if any of its records point to documents that no longer exist. The `doc_exist_cb`
+    /// callback is used to check if a document exists. It should return `true` if the document
+    /// exists and `false` otherwise.
+    ///
+    /// This function returns a delta if GC is needed, or `None` if no GC is needed.
+    pub fn scan_gc(
+        &self,
+        doc_exist_cb: fn(doc_id: t_docId) -> bool,
+    ) -> std::io::Result<Option<GcScanDelta>> {
+        let mut results = Vec::new();
+
+        for (i, block) in self.blocks.iter().enumerate() {
+            if block.num_entries == 0 {
+                results.push(BlockGcScanResult {
+                    index: i,
+                    repair: RepairType::Delete,
+                });
+                continue;
+            }
+
+            let encoder = self.encoder.clone();
+
+            let repair = block.repair(doc_exist_cb, encoder)?;
+
+            if let Some(repair) = repair {
+                results.push(BlockGcScanResult { index: i, repair });
+            }
+        }
+
+        if results.is_empty() {
+            Ok(None)
+        } else {
+            Ok(Some(GcScanDelta {
+                last_block_idx: self.blocks.len() - 1,
```

Comment:
Is an inverted index guaranteed to always have at least one block inside?

Reply:
@meiravgri yes, this is a change with the Rust implementation. The C had an extra field to keep track of the last doc ID in the index. I did not want this extra field in Rust, but rather just get it from the last block in the index. This means I cannot start with an empty block as that will cause the last doc ID for the index to be wrong.

Reply:
The extra tests are in #7088